src/monoflow.ts

```diff
@@ -17,47 +18,63 @@ export class Workflow<Z, A, B> {
     private readonly _steps: readonly Workflow<any, any, any>[] = []
   ) {}
 
-  then<C>(fn: (item: B) => C): Workflow<Z, B, C> {
+  then<C>(fn: (item: B) => C): Workflow<Z, B, C> | Workflow<Z, A, C>{
+    // Make a higher-order function when `this._err` exists.
+    if (this._ok) {
+      const binded = this._ok.bind(this);
+      return new Workflow((item: A) => fn(binded(item)), undefined, [...this._steps]);
+    }
     return new Workflow(fn, undefined, [...this._steps, this]);
   }
 
-  else<C>(fn: (error: Error) => C): Workflow<Z, B, C> {
+  else<C>(fn: (error: Error) => C): Workflow<Z, B, unknown>  {
+    // Make a higher-order function when `this._err` exists.
+    if (this._err) {
+      const binded = this._err.bind(this);
+      const safeFn = (err: Error) => {
+        try {
+          return binded(err);
+        } catch (err2) {
+          return err2;
+        }
+      }
+      return new Workflow(undefined, (error: Error) => {
+        const result = safeFn(error);
+        if (result instanceof Error) {
```

Comment:
This won't work if you throw a non-Error type:

`throw { message: "this is bad" }`

Reply:
Thanks. I completely forget about it :+1: